backend/e2e-test/routes/v3/secrets.spec.ts

```diff
@@ -0,0 +1,777 @@
+    expect(decryptedSecret.comment).toEqual(secret.comment);
+
+    // list secret should have updated value
+    const secrets = await getSecrets(seedData1.environment.slug, path);
+    expect(secrets).toEqual(
+      expect.arrayContaining([
+        expect.objectContaining({
+          key: secret.key,
+          value: "new-value",
+          type: SecretType.Shared
+        })
+      ])
+    );
+  });
+
+  test.each(testSecrets)("Delete secret in path $path", async ({ secret, path }) => {
+    const deleteSecRes = await testServer.inject({
+      method: "DELETE",
+      url: `/api/v3/secrets/${secret.key}`,
+      headers: {
+        authorization: `Bearer ${jwtAuthToken}`
+      },
+      body: {
+        workspaceId: seedData1.project.id,
+        environment: seedData1.environment.slug,
+        secretPath: path
+      }
+    });
+    expect(deleteSecRes.statusCode).toBe(200);
+    const updatedSecretPayload = JSON.parse(deleteSecRes.payload);
+    expect(updatedSecretPayload).toHaveProperty("secret");
+    const decryptedSecret = decryptSecret(projectKey, updatedSecretPayload.secret);
+    expect(decryptedSecret.key).toEqual(secret.key);
+
+    // shared secret deletion should delete personal ones also
+    const secrets = await getSecrets(seedData1.environment.slug, path);
+    expect(secrets).toEqual(
+      expect.not.arrayContaining([
+        expect.objectContaining({
+          key: secret.key,
+          type: SecretType.Shared
+        }),
+        expect.objectContaining({
+          key: secret.key,
+          type: SecretType.Personal
+        })
+      ])
+    );
+  });
+
+  test.each(testSecrets)(
+    "Deleting personal one should not delete shared secret in path $path",
+    async ({ secret, path }) => {
+      const createSharedSecRes = await testServer.inject({
+        method: "POST",
+        url: `/api/v3/secrets/${secret.key}`,
+        headers: {
+          authorization: `Bearer ${jwtAuthToken}`
+        },
+        body: {
+          workspaceId: seedData1.project.id,
+          environment: seedData1.environment.slug,
+          type: SecretType.Shared,
+          secretPath: path,
+          ...encryptSecret(projectKey, secret.key, secret.value, secret.comment)
+        }
+      });
+      expect(createSharedSecRes.statusCode).toBe(200);
+
+      const createPersonalSecRes = await testServer.inject({
+        method: "POST",
+        url: `/api/v3/secrets/${secret.key}`,
+        headers: {
+          authorization: `Bearer ${jwtAuthToken}`
+        },
+        body: {
+          workspaceId: seedData1.project.id,
+          environment: seedData1.environment.slug,
+          secretPath: path,
+          type: SecretType.Personal,
+          ...encryptSecret(projectKey, secret.key, secret.value, secret.comment)
+        }
+      });
+      expect(createPersonalSecRes.statusCode).toBe(200);
+
+      // shared secret deletion should delete personal ones also
```

Comment:
Ok this is a trade off i had made while testing. 

You can ofcourse make it independent but it doesn't make sense for integration testing. Here is why -> 

Let's say you write a test for `update service token`. But for this test to work, we need to do a `create service token` and ensure its successfully. To make this properly independent it also means we need to do a clean up. Means we need to write a function to do a delete service token and expect it to be successfully.

Now if i wrote this means i don't need to test create or delete anymore because its just redundant. Next thing is the whole test function had now all the functionality tested instead of one. 

Instead of putting everything into a test function. I split it into each into test function and inside `describe` they run sequentially. Realized we need to see it from external user perspective. Meaning how will a user delete service token. First they should be able to create one. So  the test follows same pattern.

Even if u split it the only difference would be it will be one big test function. Making it more tougher to understand

Reply:
https://github.com/Infisical/infisical/blob/abd28d92693a1d1fda01d313943d279bfa0f0093/backend/e2e-test/routes/v3/secrets.spec.ts#L239

The delete is right here

Reply:
It is generally considered bad pratice to have state in tests and it's normal for integration tests to be a bit redundant as long as we have read good readability. Here are some reasons why having state in tests is not a great idea:

- If one test fails and others depend on it, you'll have a domino effect. This makes it tough to find the root problem because you're dealing with a bunch of failures instead of just one.
- When tests are tied together, it's like untangling a big knot to figure out why something went wrong. It's confusing and will take up dev time to understand
- For a new developer, understanding a interdependent test suite can be daunting. They not only need understand what each test does individually but also how they relate to and depend on each other

Having no state between tests solve all of these issues. For example, if you need to test delete service token, you should:
- call api to create service token (what you assume to be working)
- call api to delete that token (what you are testing)

Can you make all tests independent please?

Reply:
Ok understood. Making it independent but i still don't see how this will resolve the domino effect.

Because if the creation failed this would still fail everywhere like domino effect

Reply:
You are right that related tests will fail but since each test willl be self-contained when they are independent, it will be very clear that all of those failed tests point to the same issue rather than different issues as a result of state dependence